### PR TITLE
Disallow null values from being emitted.

### LINKED ISSUE #66: NullPointerException for empty preferences.
With the change on Rx2 that `null` cannot be emitted, the simples code implementation like below crashes with NPE (stack below).

As far as I could find, it should be easy to fix.
The call that triggers a initial load on `RealPreferences@31` (`.startWith("<init>")`) should only happen `if(preferences.containsKey(key)`

```
        RxSharedPreferences rxPref = RxSharedPreferences.create(getSharedPreferences("myprefs", 0));
        Preference<String> rxString = rxPref.getString("string");
        rxString.asObservable().subscribe(new Consumer<String>() {
            @Override
            public void accept(String s) throws Exception {

            }
        });
```

```
java.lang.NullPointerException: The mapper function returned a null value.
at io.reactivex.internal.functions.ObjectHelper.requireNonNull(ObjectHelper.java:39)
at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:58)
at io.reactivex.internal.operators.observable.ObservableConcatMap$ConcatMapDelayErrorObserver.drain(ObservableConcatMap.java:462)
at io.reactivex.internal.operators.observable.ObservableConcatMap$ConcatMapDelayErrorObserver.onSubscribe(ObservableConcatMap.java:325)
at io.reactivex.internal.operators.observable.ObservableFromArray.subscribeActual(ObservableFromArray.java:29)
at io.reactivex.Observable.subscribe(Observable.java:10514)
at io.reactivex.internal.operators.observable.ObservableConcatMap.subscribeActual(ObservableConcatMap.java:54)
at io.reactivex.Observable.subscribe(Observable.java:10514)
at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
at io.reactivex.Observable.subscribe(Observable.java:10514)
at io.reactivex.Observable.subscribe(Observable.java:10500)
at io.reactivex.Observable.subscribe(Observable.java:10403)
```

### PR DESCRIPTION
**Pull request by f2prateek**
This includes two breaking changes that:

1. Remove getEnum and getObject that allowed null values.
2. Change `defaultValue` parameter from `@Nullable` to `@NonNull`.

Also makes some tools updates.

Closes #66

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#Sun Oct 02 17:31:03 PDT 2016
+#Thu Mar 09 19:25:41 PST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.3-all.zip
```
Comment (NightlyNexus) [inline gradle/wrapper/gradle-wrapper.properties]: 3.4.1 is latest.

Comment (f2prateek) [inline gradle/wrapper/gradle-wrapper.properties]: Updated

**Comment by Yukooo** [comment]
I'd leave NullPointerException because then I am aware that there is no default value and I have to deal with it instead of looking for a solution why there are no items at all (especially if there is a big chain of transformations). PS. It is like catching an exception and doing nothing then.
PS2. PLEASE DON'T ACCEPT THIS PR

**Comment by f2prateek** [comment]
@Yukooo You can still do that. This implementation doesn't hide the case when there is no null value, it simply emits the default value (which must be non-null).

**Comment by Yukooo** [comment]
@f2prateek but it is not exactly I want to see ... 
Here is my case: I have two observables one is the preference observable the other one is some default observable (it emits some default items those can have different values depending on some time period) what I do: I use onErrorResumeNext (or I would use switchIfEmpty, or default()) but NO if xxx==defaultValue in flatMap

how to check whether the value is a default? now I see only if else

**Comment by f2prateek** [comment]
Yup use an if else. Pretty easy to throw an error if you want still.

```
rxPreferences.getString("show-whats-new", DEFAULT_NULL_PREFERENCE)
    .subscribe((val) -> if (val == DEFAULT_NULL_PREFERENCE) throw new NullPointerException());
````

Not saying you should do this but the choice is yours.

**Comment by Yukooo** [comment]
hmm, as for me it is not nice since I use Preference<Dto> in the Data layer but the implementation is in App layer (Dagger2 Model where I create rxSharedPreferences.getObject(BuildConfig.KEY_DB_SOME_ITEM, null, adapter)) and in your case I have to use that DEFAULT_NULL_PREFERENCE in both layers

**Comment by Yukooo** [comment]
@f2prateek You see, I'd agree with Observable.empty() rather than with the DEFAULT_NULL_PREFERENCE

**Comment by f2prateek** [comment]
You can do the equivalent of what I suggested in your dagger provider with `map`:

```
return rxPreferences.getString("show-whats-new", DEFAULT_NULL_PREFERENCE)
    .map((val) -> {
        if (val == DEFAULT_NULL_PREFERENCE) throw new NullPointerException();
        return val;
    });
```

**Comment by Yukooo** [comment]
@f2prateek  thx, but is there any chance that you may change your mind (at least make it emits empty instead of a NonNull value)?

**Comment by f2prateek** [comment]
What is an "empty" value for objects or enums?

**Comment by Yukooo** [comment]
I mean if you don't want to pass a null reference as a default value then probably it will be better to emit nothing instead of some DEFAULT_NULL_PREFERENCE

**Comment by f2prateek** [comment]
> emit nothing instead of some DEFAULT_NULL_PREFERENCE

That's exactly the behaviour you did NOT want when you first commented.

> I'd leave NullPointerException because then I am aware that there is no default value and I have to deal with it instead of looking for a solution why there are no items at all (especially if there is a big chain of transformations).

**Comment by Yukooo** [comment]
you is right, but we are all people, so if I had to choose DEFAULT or EMPTY I'd like to choose EMPTY

**Comment by ZakTaccardi** [comment]
The exception should not be thrown because it's a terminal event - but we still need a way of easily observing a hot stream of a preference value.

**Pull request merged**